### PR TITLE
The final arrival of Serving size and Cookware on KooCook!

### PR DESCRIPTION
**Pull request by th-bunratta**
This branch introduces these features:

- Recipe yield unit customisation
- Dynamic unit conversion
- Equipment & Cookware information in Recipes 
  - KooCook users can now associate cookware with recipes
- Recipe portal navigation bar

Follow-up features:
- Minor colour theming and decorations on the site

Any assistance with UI design would be grateful.

### CODE REVIEW

File: koocook_core/models/base.py [hunk 1]
```diff
@@ -31,7 +31,10 @@ def as_json(self) -> str:
...
     def default(self, obj: models.Model):
         if hasattr(obj, 'as_dict'):
-            return obj.as_dict()
+            try:
+                return obj.as_dict()
+            except TypeError:
+                return obj.as_dict()
         else:
             if isinstance(obj, models.Model):
                 if isinstance(obj, SerialisableModel):
```
Comment (MaiMee1) [inline koocook_core/models/base.py]: How does this work? Wouldn't doing the same statement raise the same error?

File: koocook_core/support/query/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from .ruleset import *
```
Comment (MaiMee1) [inline koocook_core/support/query/__init__.py]: Missing a LF/CRLF

File: koocook_core/support/query/ruleset.py [hunk 3]
```diff
@@ -1,21 +1,68 @@
...
+        elif request_fields.get(self.key) and request_fields.get(self.key) != '':
+            self.rule_body = request_fields.get(self.key)
+            return self.process(request_fields, queryset)
+        else:
+            return queryset
+
+
+class QueryRuleset:
+    def __init__(self, *args: Type[Rule]):
+        self.rules: Tuple[Type[Rule]] = args
```
Comment (MaiMee1) [inline koocook_core/support/query/ruleset.py]: You sure it's `Tuple[Type[Rule]]` for tuple of length one and not `Tuple[Type[Rule], ...]` of variable length?

File: koocook_core/urls/__init__.py [hunk 4]
```diff
@@ -16,7 +16,13 @@
...
+    path('cookware/', views.RecipeEquipmentView.as_view(), name='equipment-all'),
+    path('cookware/<int:pk>', views.RecipeEquipmentDetailView.as_view(), name='equipment'),
+    path('ingredients/', views.RecipeIngredientsView.as_view(), name='ingredients'),
+    path('ingredients/<int:pk>', views.IngredientDetailView.as_view(), name='ingredient'),
+    path('recipes/', include('koocook_core.urls.recipes'), name='recipes'),
     path('recipes/', views.PreferredRecipeStreamView.as_view(), name='recipe-all'),
+    path('recipes/unit_conv', recipe_handler.handle, name='recipe-unit-conv', kwargs={"alias": 'unit_conv'}),
     path('recipes/<int:recipe_id>', views.handle_recipe, name='recipe'),
     path('recipes/<int:item_id>/comments', recipe_handler.handle, name='recipe-comments', kwargs={"alias": 'comment'}),
     path('recipes/<int:pk>/edit', views.RecipeUpdateView.as_view(), name='recipe-edit'),
```
Comment (MaiMee1) [inline koocook_core/urls/__init__.py]: Why the redundancy?

Comment (th-bunratta) [reply koocook_core/urls/__init__.py]: Legacy code. Removal pending

**Comment by MaiMee1** [comment]
When you go into search `/search/?order=name`, for example, "Failed to follow! Try again" pops up unexpectedly. Why is that?

File: koocook_core/support/query/__init__.py (hunk 2 above)
Comment (MaiMee1) [reply koocook_core/support/query/__init__.py]: Resolved in 3f5b9074e1ced672d0924e7327f6287671517f17

File: koocook_core/support/query/ruleset.py (hunk 3 above)
Comment (th-bunratta) [reply koocook_core/support/query/ruleset.py]: Does that affect the way type checking behave?

File: koocook_core/models/base.py (hunk 1 above)
Comment (th-bunratta) [reply koocook_core/models/base.py]: Fixed. Thanks!

**Comment by th-bunratta** [comment]
> When you go into search `/search/?order=name`, for example, "Failed to follow! Try again" pops up unexpectedly. Why is that?

Try hard resetting the page on your browser, I guess?

**Comment by MaiMee1** [comment]
So it's not a bug? `/search/` with no sign in also results in this error

**Comment by th-bunratta** [comment]
> So it's not a bug? `/search/` with no sign in also results in this error

But that causes a lil' frustration to me. I'll fix it.

File: koocook_core/urls/__init__.py (hunk 4 above)
Comment (MaiMee1) [reply koocook_core/urls/__init__.py]: Fixed in cf75dd5f99e1a8d0c1eced6fb56c6d1ef9a38542 for you

File: koocook_core/support/query/ruleset.py (hunk 3 above)
Comment (MaiMee1) [reply koocook_core/support/query/ruleset.py]: Yes it does

**Comment by th-bunratta** [comment]
> I helped reconfigure the URLs for you. I'm not the author so I'm not sure if I missed anything. Try some manual tests before merging!

Huge thanks!

**Comment by MaiMee1** [comment]
Some things highlighted by [Codacy](https://app.codacy.com/manual/KooCook/koocook-dj/pullRequest?prid=4673451) you might want to look into
image
image

**Pull request merged**